### PR TITLE
Capture of scenarios motivating feature testing.

### PR DESCRIPTION
**Pull request by mtrofin**
Moved a few of the already described scenarios, trying to isolate the motivating problem from possible solutions.

Added a development time scenario and one for restrictive target environment.

### CODE REVIEW

File: FeatureTest.md [hunk 1]
```diff
@@ -1,3 +1,18 @@
+# Motivating Scenarios
+1. In MVP + 1 (the version of the WebAssembly after MVP), [`i32.min_s`](FutureFeatures.md#additional-integer-operations) is introduced. A developer updates her toolkit and targets the new features. The compiler leverages `i32.min_s` when compiling a module the developer wrote. The developer's WebAssembly module works correctly both on execution environments at MVP, as well as those updated on MVP + 1.
```
Comment (jfbastien) [inline FeatureTest.md]: We call this [Essential post-MVP features](PostMVP.md).

Comment (jfbastien) [inline FeatureTest.md]: I think we've tried going gender-neutral everywhere, so "developers update their".

Comment (mtrofin) [inline FeatureTest.md]: OK - correcting.

File: FeatureTest.md [hunk 2]
```diff
@@ -1,3 +1,18 @@
+# Motivating Scenarios
+1. In MVP + 1 (the version of the WebAssembly [after MVP](PostMVP.md)), [`i32.min_s`](FutureFeatures.md#additional-integer-operations) is introduced. A WebAssembly developer updates their toolkit and targets the new features. The compiler leverages `i32.min_s` when compiling a module the developer wrote. The developer's WebAssembly module works correctly both on execution environments at MVP, as well as those updated on MVP + 1.
```
Comment (jfbastien) [inline FeatureTest.md]: I'd drop "MVP +1" since that's what PostMVP is. Just call it PostMVP.

File: FeatureTest.md [hunk 3]
```diff
@@ -1,3 +1,18 @@
...
+
+2. In MVP + 1, module authors may now use [Threading](PostMVP.md#threads) APIs in the browser. A developer wants to leverage multithreading in their module.
+  * In one variant of the scenario, our developer does not want to pay the engineering cost of developing and supporting a threaded and non-threaded version of their code. They opt not to support MVP targets, and only support MVP + 1 targets. End-users (browser users) get some message indicating they need MVP support.
+  * In another variant, our developer explicitly authors both MVP-only and MVP + 1 (with threads) code.
```
Comment (jfbastien) [inline FeatureTest.md]: One thing that's not very clear throughout the examples here: MVP is the only "version" of WebAssembly. After MVP it's features which implementors can ship asynchronously. There's no "version 2" planned (we may do one, but it's not in the plans).

That means that feature testing is pretty basic in the model we assume: you can't check the version, all you have are feature tests!

File: FeatureTest.md [hunk 4]
```diff
@@ -1,3 +1,18 @@
...
+  
+3. [SIMD](PostMVP.md#fixed-width-simd) support is not universally equivalent on all targets. While polyfill variants of SIMD APIs are available, a developer prefers writing dedicated SIMD and non-SIMD versions of their compresion algorithm, because the non-SIMD version performs better in environments without SIMD support, when compared to the SIMD polyfill. They package their compression code for reuse by third parties.
+
+4. An application author is assembling together an application by reusing modules such as those developed in the scenarios above. The application author's development environment is able to quickly and correctly identify the platform dependencies (e.g. threading, SIMD) and communicate back to the application author the implications these dependencies have on the end-application. Some APIs exposed from the threading-aware module are only pertinent to environments supporting threading. As a consequence, the application author needs to write specialized code when threads are/are not supported.
+(Note: we should understand ghis scenario for both forms of WebAssembly reuse currently imagined - dynamic linking and static imports.)
```
Comment (jfbastien) [inline FeatureTest.md]: Typo "ghis"

Comment (jfbastien) [inline FeatureTest.md]: Use em dash.

File: FeatureTest.md [hunk 5]
```diff
@@ -1,3 +1,18 @@
...
+
+5. Jane's compression algorithm module is deployed on a restrictive execution environment, as part of an application. In this environment, a process may not change memory page access protection flags (e.g. certain gaming consoles, to investigate server side deployment scenarios). Jane's module is compiled by the WebAssembly environment, enabling the configuration most specific to the target (i.e. with/without Threads, SIMD, etc).
```
Comment (jfbastien) [inline FeatureTest.md]: Jane?

**Comment by jfbastien** [comment]
As discussed offline, maybe move this to Rationale.md?

File: FeatureTest.md (hunk 3 above)
Comment (mtrofin) [inline FeatureTest.md]: I see. I updated referencing PostMVP as such, but not assuming anything about whether there will or not be a coarser grained "WebAssembly v2" milestone, or just finer grained features. The scenarios should help us tease apart what makes sense here.

File: FeatureTest.md (hunk 4 above)
Comment (mtrofin) [inline FeatureTest.md]: What's em dash?

**Comment by mtrofin** [comment]
- moved the scenarios to Rationale.md, in their own section
- referenced the section in FeatureTest.md
- incorporated feedback

File: Rationale.md [hunk 6]
```diff
@@ -240,3 +240,20 @@ architectures there may be a need to revisit some of the decisions:
 * When different languages have different expectations then it's unfortunate if
   WebAssembly measurably penalizes one's performance by enforcing determinism
   which that language doesn't care about, but which another language may want.
+
+
+## Feature Testing - Motivating Scenarios
```
Comment (jfbastien) [inline Rationale.md]: "Motivating Scenarios for Feature Testing"? Different Markdown parsers treat dashes differently (should be em dash too :p)

File: Rationale.md [hunk 7]
```diff
@@ -240,3 +240,20 @@ architectures there may be a need to revisit some of the decisions:
...
+1. [PostMVP](PostMVP.md), [`i32.min_s`](FutureFeatures.md#additional-integer-operations) is introduced. A WebAssembly developer updates their toolkit so that the compiler may leverage `i32.min_s`. The developer's WebAssembly module works correctly both on execution environments at MVP, as well as those supporting `i32.min_s`.
```
Comment (jfbastien) [inline Rationale.md]: The link in brackets can just be "post-MVP".

File: Rationale.md [hunk 8]
```diff
@@ -240,3 +240,20 @@ architectures there may be a need to revisit some of the decisions:
...
+  * A variant of this, where a few more new opcodes are available, the compiler is updated to be able to leverage all of them, but not all execution targets support all of them. The developer wants to reach as many of their customers as possible, while at the same time providing them with the best experience possible. The developer has to balance the cost of the test matrix resulting from the combinations of possible feature configurations.
+
+2. PostMVP, module authors may now use [Threading](PostMVP.md#threads) APIs in the browser. A developer wants to leverage multithreading in their module.
+  * In one variant of the scenario, our developer does not want to pay the engineering cost of developing and supporting a threaded and non-threaded version of their code. They opt not to support MVP targets, and only support MVP + 1 targets. End-users (browser users) get some message indicating they need MVP support.
```
Comment (jfbastien) [inline Rationale.md]: "MVP + 1" here.

File: Rationale.md [hunk 9]
```diff
@@ -240,3 +240,20 @@ architectures there may be a need to revisit some of the decisions:
...
+  * In another variant, our developer explicitly authors both MVP-only and MVP + 1 (with threads) code.
```
Comment (jfbastien) [inline Rationale.md]: Same.

File: Rationale.md [hunk 10]
```diff
@@ -240,3 +240,20 @@ architectures there may be a need to revisit some of the decisions:
...
+  
+3. [SIMD](PostMVP.md#fixed-width-simd) support is not universally equivalent on all targets. While polyfill variants of SIMD APIs are available, a developer prefers writing dedicated SIMD and non-SIMD versions of their compresion algorithm, because the non-SIMD version performs better in environments without SIMD support, when compared to the SIMD polyfill. They package their compression code for reuse by third parties.
```
Comment (jfbastien) [inline Rationale.md]: "compression"

File: Rationale.md [hunk 11]
```diff
@@ -240,3 +240,20 @@ architectures there may be a need to revisit some of the decisions:
...
+
+4. An application author is assembling together an application by reusing modules such as those developed in the scenarios above. The application author's development environment is able to quickly and correctly identify the platform dependencies (e.g. threading, SIMD) and communicate back to the application author the implications these dependencies have on the end-application. Some APIs exposed from the threading-aware module are only pertinent to environments supporting threading. As a consequence, the application author needs to write specialized code when threads are/are not supported.
+(Note: we should understand this scenario for both forms of WebAssembly reuse currently imagined - dynamic linking and static imports.)
```
Comment (jfbastien) [inline Rationale.md]: Em dash still.

File: Rationale.md [hunk 12]
```diff
@@ -240,3 +240,20 @@ architectures there may be a need to revisit some of the decisions:
...
+
+5. The compression algorithm described in scenario 3 is deployed on a restrictive execution environment, as part of an application. In this environment, a process may not change memory page access protection flags (e.g. certain gaming consoles, to investigate server side deployment scenarios). Jane's module is compiled by the WebAssembly environment, enabling the configuration most specific to the target (i.e. with/without Threads, SIMD, etc).
```
Comment (jfbastien) [inline Rationale.md]: Jane is still here.

File: Rationale.md (hunk 11 above)
Comment (mtrofin) [inline Rationale.md]: Assuming "em-dash" means that longer dash, I couldn't figure out how. Docs say "---". Nope. 

I changed it to ":"

**Comment by mtrofin** [comment]
Updated, and avoided "em-dash"-es, since I can't figure out how to make them.

File: Rationale.md [hunk 13]
```diff
@@ -240,3 +240,20 @@ architectures there may be a need to revisit some of the decisions:
...
   WebAssembly measurably penalizes one's performance by enforcing determinism
   which that language doesn't care about, but which another language may want.
+
+
+## Motivating Scenarios for Feature Testing
+1. [Post-MVP](PostMVP.md), [`i32.min_s`](FutureFeatures.md#additional-integer-operations) is introduced. A WebAssembly developer updates their toolkit so that the compiler may leverage `i32.min_s`. The developer's WebAssembly module works correctly both on execution environments at MVP, as well as those supporting `i32.min_s`.
+  * A variant of this, where a few more new opcodes are available, the compiler is updated to be able to leverage all of them, but not all execution targets support all of them. The developer wants to reach as many of their customers as possible, while at the same time providing them with the best experience possible. The developer has to balance the cost of the test matrix resulting from the combinations of possible feature configurations.
+
+2. PostMVP, module authors may now use [Threading](PostMVP.md#threads) APIs in the browser. A developer wants to leverage multithreading in their module.
```
Comment (jfbastien) [inline Rationale.md]: "Post-MVP" here too.

**Comment by jfbastien** [comment]
lgtm after the last comment.

Em dash is Unicode, you can just use it: —

**Comment by mtrofin** [comment]
Done! Thanks.

**Pull request merged**